inode/inode.go

```diff
@@ -373,55 +373,52 @@ func (vS *volumeStruct) inodeCacheDiscard() {
 		inodesToDrop = (vS.inodeCacheLRUItems * globals.inodeSize) - vS.inodeCacheLRUMaxBytes
 		inodesToDrop = inodesToDrop / globals.inodeSize
 		inodesToDrop += inodesToDrop / 4
-	}
-	for ((vS.inodeCacheLRUItems * globals.inodeSize) > vS.inodeCacheLRUMaxBytes) &&
-		(inodesToDrop > 0) {
-		inodesToDrop--
+		for ((vS.inodeCacheLRUItems * globals.inodeSize) > vS.inodeCacheLRUMaxBytes) &&
+			(inodesToDrop > 0) {
+			inodesToDrop--
+
+			ic := vS.inodeCacheLRUHead
+
+			// Create a DLM lock object
+			id := dlm.GenerateCallerID()
+			inodeRWLock, _ := vS.InitInodeLock(ic.InodeNumber, id)
+			err := inodeRWLock.TryWriteLock()
+
+			// Inode is locked; skip it
+			if err != nil {
+				// Move inode to tail of LRU
+				vS.inodeCacheTouchWhileLocked(ic)
+				continue
+			}
 
-		ic := vS.inodeCacheLRUHead
+			if ic.dirty {
+				// The inode is busy - drop the DLM lock and move to tail
+				inodeRWLock.Unlock()
+				vS.inodeCacheTouchWhileLocked(ic)
+				continue
+			}
 
-		// Create a DLM lock object
-		id := dlm.GenerateCallerID()
-		inodeRWLock, _ := vS.InitInodeLock(ic.InodeNumber, id)
-		err := inodeRWLock.TryWriteLock()
+			var ok bool
 
-		// Inode is locked; skip it
-		if err != nil {
-			// Move inode to tail of LRU
-			vS.inodeCacheTouchWhileLocked(ic)
-			continue
-		}
+			ok, err = vS.inodeCacheDropWhileLocked(ic)
+			if err != nil || !ok {
+				pStr := fmt.Errorf("The inodes was not found in the inode cache - ok: %v err: %v", ok, err)
+				panic(pStr)
+			}
 
-		if (ic.inFlightLogSegmentMap != nil) && (ic.dirty == false) {
-			// The inode is busy - drop the DLM lock and move to tail
+			// NOTE: Releasing the locks out of order here.
+			//
+			// We acquire the locks in this order:
+			// 1. Volume lock
+			// 2. DLM lock for inode
+			//
+			// We then release the volume lock before deleting the inode from the cache and
+			// then releasing the DLM lock.
 			inodeRWLock.Unlock()
-			vS.inodeCacheTouchWhileLocked(ic)
-			continue
-		}
-
-		vS.Unlock()
-
-		var ok bool
 
-		ok, err = vS.inodeCacheDrop(ic)
-		if err != nil || !ok {
-			pStr := fmt.Errorf("The inodes was not found in the inode cache - ok: %v err: %v", ok, err)
-			panic(pStr)
+			// NOTE: vS.inodeCacheDropWhileLocked() removed the inode from the LRU list so
+			// the head is now different
 		}
-
-		// NOTE: Releasing the locks out of order here.
-		//
-		// We acquire the locks in this order:
-		// 1. Volume lock
-		// 2. DLM lock for inode
-		//
-		// We then release the volume lock before deleting the inode from the cache and
-		// then releasing the DLM lock.
-		inodeRWLock.Unlock()
-
-		// vS.inodeCacheDrop() removed the inode from the freelist so
-		// the head is now different
-		vS.Lock()
 	}
 	vS.Unlock()
```

Comment:
So, just to be clear, we will hold the vS.Lock while in the loop?

If that is so, then I think the test at the top of the loop can be simplified to just `inodesToDrop > 0` since it can only be "this loop" that decrements vS.inodeCacheLRUItems, right? If two such timer tick instances happen to be running at the same time, only one of them will have the vS.Lock and the other will be blocked... and presumably find nothing to do when they eventually get the lock.